type.go

```diff
@@ -1693,83 +1841,59 @@ func (t *timeType) ConvertedType() *deprecated.ConvertedType {
-	} else {
-		return newInt64ColumnBuffer(t, makeColumnIndex(columnIndex), makeNumValues(numValues))
-	}
+	return t.baseType().NewColumnBuffer(columnIndex, numValues)
 }
 
 func (t *timeType) NewDictionary(columnIndex, numValues int, data encoding.Values) Dictionary {
-	if t.useInt32() {
-		return newInt32Dictionary(t, makeColumnIndex(columnIndex), makeNumValues(numValues), data)
-	} else {
-		return newInt64Dictionary(t, makeColumnIndex(columnIndex), makeNumValues(numValues), data)
-	}
+	return t.baseType().NewDictionary(columnIndex, numValues, data)
 }
 
 func (t *timeType) NewPage(columnIndex, numValues int, data encoding.Values) Page {
-	if t.useInt32() {
-		return newInt32Page(t, makeColumnIndex(columnIndex), makeNumValues(numValues), data)
-	} else {
-		return newInt64Page(t, makeColumnIndex(columnIndex), makeNumValues(numValues), data)
-	}
+	return t.baseType().NewPage(columnIndex, numValues, data)
 }
 
-func (t *timeType) NewValues(values []byte, _ []uint32) encoding.Values {
-	if t.useInt32() {
-		return encoding.Int32ValuesFromBytes(values)
-	} else {
-		return encoding.Int64ValuesFromBytes(values)
-	}
+func (t *timeType) NewValues(values []byte, offset []uint32) encoding.Values {
+	return t.baseType().NewValues(values, offset)
 }
 
 func (t *timeType) Encode(dst []byte, src encoding.Values, enc encoding.Encoding) ([]byte, error) {
-	if t.useInt32() {
-		return encoding.EncodeInt32(dst, src, enc)
-	} else {
-		return encoding.EncodeInt64(dst, src, enc)
-	}
+	return t.baseType().Encode(dst, src, enc)
 }
 
 func (t *timeType) Decode(dst encoding.Values, src []byte, enc encoding.Encoding) (encoding.Values, error) {
-	if t.useInt32() {
-		return encoding.DecodeInt32(dst, src, enc)
-	} else {
-		return encoding.DecodeInt64(dst, src, enc)
-	}
+	return t.baseType().Decode(dst, src, enc)
 }
 
 func (t *timeType) EstimateDecodeSize(numValues int, src []byte, enc encoding.Encoding) int {
-	if t.useInt32() {
-		return Int32Type.EstimateDecodeSize(numValues, src, enc)
-	} else {
-		return Int64Type.EstimateDecodeSize(numValues, src, enc)
-	}
+	return t.baseType().EstimateDecodeSize(numValues, src, enc)
 }
 
 func (t *timeType) AssignValue(dst reflect.Value, src Value) error {
-	if t.useInt32() {
-		return Int32Type.AssignValue(dst, src)
-	} else {
-		return Int64Type.AssignValue(dst, src)
-	}
+	return t.baseType().AssignValue(dst, src)
 }
 
 func (t *timeType) ConvertValue(val Value, typ Type) (Value, error) {
-	if t.useInt32() {
-		return Int32Type.ConvertValue(val, typ)
-	} else {
-		return Int64Type.ConvertValue(val, typ)
+	switch src := typ.(type) {
+	case *stringType:
+		tz := t.tz()
+		if t.Unit.Micros != nil {
+			return convertStringToTimeMicros(val, tz)
+		} else {
+			return convertStringToTimeMillis(val, tz)
+		}
+	case *timestampType:
+		tz := t.tz()
+		if t.Unit.Micros != nil {
+			return convertTimestampToTimeMicros(val, src.Unit, src.tz(), tz)
+		} else {
+			return convertTimestampToTimeMillis(val, src.Unit, src.tz(), tz)
```

Comment:
Maybe just a style thing but I always prefer 

```go
if t.Unit.Micros != nil {
			return convertTimestampToTimeMicros(val, src.Unit, src.tz(), tz)
		}
                 return convertTimestampToTimeMillis(val, src.Unit, src.tz(), tz)
```

which saves you one line and a layer of indent.

Reply:
I think the alignment of statements help make sense of the code, but that's obviously a very personal opinion.

Let's not aim to follow conventions that a formatting tool isn't going to enforce for us, we'll waste too much time doing a job that a machine would be much better than us at 👍